include/aws/crt/Variant.h

```diff
@@ -0,0 +1,348 @@
+//                template<typename T, typename First, typename Second, typename ...Rest>
+//                constexpr bool ContainsType() {
+//                    return std::is_same<T, First>::value || ContainsType<T, Second, Rest...>();
+//                }
+                template<typename T, typename Last>
+                constexpr bool ContainsType() {
+                    return std::is_same<T, Last>::value;
+                }
+
+                template<typename T, typename First, typename Second, typename ...Rest>
+                constexpr bool ContainsType() {
+                    return std::is_same<T, First>::value || ContainsType<T, Second, Rest...>();
+                }
+
+//                template<typename T, typename ...Ts>
+//                constexpr bool ContainsType() {
+//                    return ContainsType<T, Ts...>();
+//                }
+
+
+
+                // a case when the template parameter pack is empty (i.e. Variant<>)
+                template<typename T>
+                constexpr bool ContainsType() {
+                    return false;
+                }
+
+                template<typename T, typename ...Ts>
+                struct HasType
+                {
+                    static const bool value = ContainsType<T, Ts...>();
+                };
+            }
+        }
+
+        template <std::size_t Index, typename... Ts> class VariantAlternative;
+
+        /**
+         * Custom implementation of a Variant type. std::variant requires C++17
+         * @tparam Ts types of the variant value
+         */
+        template <typename... Ts> class Variant
+        {
+          public:
+            static constexpr std::size_t AlternativeCount = sizeof...(Ts);
+
+            template<std::size_t Index>
+            using VariantAlternative = VariantAlternative<Index, Variant<Ts...>>;
+
+            using ThisVariantT = Variant<Ts...>;
+            template<typename OtherT>
+            using EnableIfOtherIsSameVariantType = typename std::enable_if<std::is_same<typename std::decay<OtherT>, ThisVariantT>::value, int>::type;
+
+            template<typename OtherT>
+            using EnableIfOtherIsThisVariantAlternative = typename std::enable_if<VariantDetail::Checker::HasType<typename std::decay<OtherT>::type, Ts...>::value, int>::type;
+
+            Variant() : m_value(nullptr) {}
+
+            // SFINAE to force compiler to choose a correct overload between Variant(anotherVariant) and Variant(alternative) constructors
+            template<typename VariantT = ThisVariantT, EnableIfOtherIsSameVariantType<VariantT> = 1>
+            Variant(const VariantT &other)
+            {
+                auto otherIdx = other.m_index;
+                if (otherIdx != -1)
+                {
+                    CopyAssignUtil(other);
+                }
+            }
+
+            template<typename VariantT = ThisVariantT, EnableIfOtherIsSameVariantType<VariantT> = 1>
+            Variant(VariantT &&other)
```

Comment:
should we consider using universal references? seems like a lot of this is split between a rvalue and a lvalue path. I too dont love universal refs but it feels like it could be used here